File: ios/App/App/WidgetBridgePlugin.swift
import Foundation
import Capacitor
#if canImport(WidgetKit)
import WidgetKit
#endif

@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetBridgePlugin"
    public let jsName = "WidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setNextPrayer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAyahOfDay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reloadAll", returnType: CAPPluginReturnPromise)
    ]

    /// Must match the App Group enabled on BOTH the App and the Widget Extension targets.
    /// Add it via Xcode → target → Signing & Capabilities → App Groups.
    public static let appGroup = "group.app.quran.fady"

    public enum Keys {
        public static let nextPrayerName = "nextPrayerName"
        public static let nextPrayerNameAr = "nextPrayerNameAr"
        public static let nextPrayerTime = "nextPrayerTime"
        public static let nextPrayerRemainingMinutes = "nextPrayerRemainingMinutes"
        public static let nextPrayerLocation = "nextPrayerLocation"
        public static let nextPrayerIsTomorrow = "nextPrayerIsTomorrow"
        public static let nextPrayerUpdatedAt = "nextPrayerUpdatedAt"

        public static let ayahText = "ayahText"
        public static let ayahSurah = "ayahSurah"
        public static let ayahNumber = "ayahNumber"
        public static let ayahUpdatedAt = "ayahUpdatedAt"
    }

    private func defaults() -> UserDefaults? {
        return UserDefaults(suiteName: WidgetBridgePlugin.appGroup)
    }

    @objc func setNextPrayer(_ call: CAPPluginCall) {
        guard let d = defaults() else {
            call.reject("App Group '\(WidgetBridgePlugin.appGroup)' is not configured. Enable it on the App and Widget targets.")
            return
        }
        d.set(call.getString("name") ?? "", forKey: Keys.nextPrayerName)
        d.set(call.getString("nameAr") ?? "", forKey: Keys.nextPrayerNameAr)
        d.set(call.getString("time") ?? "", forKey: Keys.nextPrayerTime)
        d.set(call.getInt("remainingMinutes") ?? 0, forKey: Keys.nextPrayerRemainingMinutes)
        d.set(call.getString("location") ?? "", forKey: Keys.nextPrayerLocation)
        d.set(call.getBool("isTomorrow") ?? false, forKey: Keys.nextPrayerIsTomorrow)
        d.set(Date().timeIntervalSince1970, forKey: Keys.nextPrayerUpdatedAt)

        reloadTimelines()
        call.resolve()
    }

    @objc func setAyahOfDay(_ call: CAPPluginCall) {
        guard let d = defaults() else {
            call.reject("App Group '\(WidgetBridgePlugin.appGroup)' is not configured.")
            return
        }
        d.set(call.getString("text") ?? "", forKey: Keys.ayahText)
        d.set(call.getString("surah") ?? "", forKey: Keys.ayahSurah)
        d.set(call.getInt("ayah") ?? 0, forKey: Keys.ayahNumber)
        d.set(Date().timeIntervalSince1970, forKey: Keys.ayahUpdatedAt)

        reloadTimelines()
        call.resolve()
    }

    @objc func reloadAll(_ call: CAPPluginCall) {
        reloadTimelines()
        call.resolve()
    }

    private func reloadTimelines() {
        #if canImport(WidgetKit)
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        #endif
    }
}

File: android/app/src/main/java/app/quran/fady/widgets/WidgetBridgePlugin.java
package app.quran.fady.widgets;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.SharedPreferences;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {

    /** Must match the SharedPreferences name used by the AppWidget providers. */
    public static final String PREFS_NAME = "QuranWidgetPrefs";

    public static final String KEY_PRAYER_NAME = "nextPrayerName";
    public static final String KEY_PRAYER_NAME_AR = "nextPrayerNameAr";
    public static final String KEY_PRAYER_TIME = "nextPrayerTime";
    public static final String KEY_PRAYER_REMAINING = "nextPrayerRemainingMinutes";
    public static final String KEY_PRAYER_LOCATION = "nextPrayerLocation";
    public static final String KEY_PRAYER_IS_TOMORROW = "nextPrayerIsTomorrow";
    public static final String KEY_PRAYER_UPDATED_AT = "nextPrayerUpdatedAt";

    public static final String KEY_AYAH_TEXT = "ayahText";
    public static final String KEY_AYAH_SURAH = "ayahSurah";
    public static final String KEY_AYAH_NUMBER = "ayahNumber";
    public static final String KEY_AYAH_UPDATED_AT = "ayahUpdatedAt";

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    @PluginMethod
    public void setNextPrayer(PluginCall call) {
        SharedPreferences.Editor e = prefs().edit();
        e.putString(KEY_PRAYER_NAME, call.getString("name", ""));
        e.putString(KEY_PRAYER_NAME_AR, call.getString("nameAr", ""));
        e.putString(KEY_PRAYER_TIME, call.getString("time", ""));
        e.putInt(KEY_PRAYER_REMAINING, call.getInt("remainingMinutes", 0));
        e.putString(KEY_PRAYER_LOCATION, call.getString("location", ""));
        e.putBoolean(KEY_PRAYER_IS_TOMORROW, call.getBoolean("isTomorrow", false));
        e.putLong(KEY_PRAYER_UPDATED_AT, System.currentTimeMillis());
        e.apply();

        refreshWidgets();
        call.resolve();
    }

    @PluginMethod
    public void setAyahOfDay(PluginCall call) {
        SharedPreferences.Editor e = prefs().edit();
        e.putString(KEY_AYAH_TEXT, call.getString("text", ""));
        e.putString(KEY_AYAH_SURAH, call.getString("surah", ""));
        e.putInt(KEY_AYAH_NUMBER, call.getInt("ayah", 0));
        e.putLong(KEY_AYAH_UPDATED_AT, System.currentTimeMillis());
        e.apply();

        refreshWidgets();
        call.resolve();
    }

    @PluginMethod
    public void reloadAll(PluginCall call) {
        refreshWidgets();
        call.resolve();
    }

    private void refreshWidgets() {
        Context ctx = getContext();
        AppWidgetManager mgr = AppWidgetManager.getInstance(ctx);

        for (Class<?> cls : new Class<?>[] { NextPrayerWidgetProvider.class, AyahWidgetProvider.class }) {
            ComponentName cn = new ComponentName(ctx, cls);
            int[] ids = mgr.getAppWidgetIds(cn);
            if (ids != null && ids.length > 0) {
                android.content.Intent intent = new android.content.Intent(ctx, cls);
                intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
                intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
                ctx.sendBroadcast(intent);
            }
        }
    }
}
